lib/proxymode.pm

```diff
@@ -0,0 +1,182 @@
+# SUSE's openQA tests
+#
+# Copyright © 2012-2016 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+#
+package proxymode;
+# Summary: proxymode: The basic lib for using proxy mode to connect or operation with physcial machine
+# Maintainer: John <xgwang@suse.com>
+
+use base 'basetest';
+use testapi;
+use strict;
+
+our $SLAVE_SERIALDEV = 'proxyserial';
+our $DEBUG           = 1;
+
+sub restart_host {
+    my $self         = shift;
+    my $ipmi_machine = shift;
+    my $ipmi_user    = shift;
+    my $ipmi_pass    = shift;
+
+    if (!$ipmi_pass) {
+        $ipmi_pass = 'ADMIN';
+    }
+    if (!$ipmi_user) {
+        $ipmi_user = 'ADMIN';
+    }
+    if (!$ipmi_machine) {
+        die "Variable ipmi_machine is invalid in function restart_host!!\n";
+    }
+
+    my $ipmitool = "ipmitool -H " . $ipmi_machine . " -U " . $ipmi_user . " -P " . $ipmi_pass . " -I lanplus ";
+
+    select_console 'log-console';
+    script_run($ipmitool . 'chassis power off', 20);
+    while (1) {
+        my $stdout = script_output($ipmitool . 'chassis power status', 20);
+        last if $stdout =~ m/is off/;
+        die "Failure on running IPMITOOL" if $stdout =~ m/Error/;
+        script_run("", $ipmitool . 'chassis power off', 20);
+        sleep(2);
+    }
+    script_run($ipmitool . 'chassis power on', 20);
+    while (1) {
+        my $ret = script_output($ipmitool . 'chassis power status', 20);
+        last if $ret =~ m/is on/;
+        script_run("", $ipmitool . 'chassis power on', 20);
+        sleep(2);
+    }
+    wait_idle 10;
+    save_screenshot;
+    select_console 'root-console';
+}
+
+sub connect_slave {
+    my $self         = shift;
+    my $ipmi_machine = shift;
+    my $ipmi_user    = shift;
+    my $ipmi_pass    = shift;
+    if (!$ipmi_user) {
+        $ipmi_user = 'ADMIN';
+    }
+    if (!$ipmi_pass) {
+        $ipmi_pass = 'ADMIN';
+    }
+    if (!$ipmi_machine) {
+        die "Variable ipmi_machine is invalid in function connect_slave!!\n";
+    }
+    script_run("clear");
+    type_string("ipmitool -H " . $ipmi_machine . " -U " . $ipmi_user . " -P " . $ipmi_pass . " -I lanplus sol activate", 20);
+    send_key 'ret';
+    send_key 'ret';
+    save_screenshot;
+}
+
+sub check_prompt_for_boot() {
+    my $self    = shift;
+    my $timeout = shift;
+    if (!$timeout) {
+        $timeout = 5000;
+    }
+    assert_screen("autoyast-system-login-console", $timeout);
+    type_string "root\n";
+    sleep 2;
+    type_password;
+    send_key "ret";
+    assert_screen("text-logged-in-root", 10);
+    type_string("clear;ip a\n");
+}
+
+sub save_org_serialdev() {
+    my $self = shift;
+    if (!get_var("PROXY_SERIALDEV")) {
+        set_var("PROXY_SERIALDEV", $serialdev);
```

Comment:
you shouldn't need this if you use `get_var` with the default parameter, e.g.

```perl
get_var('PROXY_SERIALDEV', $serialdev)
```

will either use the overridden value of 'PROXY_SERIALDEV' or revert to the default `$serialdev`